libs/traits/src/investments.rs

```diff
@@ -0,0 +1,417 @@
+// Copyright 2021 Centrifuge GmbH (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use sp_std::fmt::Debug;
+
+/// A trait for converting from a PoolId and a TranchId
+/// into a given Self::Currency
+pub trait TrancheCurrency<PoolId, TrancheId> {
+	fn generate(pool_id: PoolId, tranche_id: TrancheId) -> Self;
+
+	fn of_pool(&self) -> PoolId;
+
+	fn of_tranche(&self) -> TrancheId;
+}
+
+/// A trait, when implemented allows to invest into
+/// investment classes
+pub trait Investment<AccountId> {
```

Comment:
It will probably be a future modification if all of you agree. `Investment` and `ForeingInvestment` have an exact parallelism regarding investment & redeemption, so maybe that parallelism can be extracted as follows:

```rust
enum InvestDirection {
    In, //which means investment
    Out, //which means redemption
}

trait Investment<InvestDirection> {
    // Just 4 methods instead of 8
    fn update();
    fn get(); // previous investment() & redemption()
    fn accepted_currency();
    fn requires_collect();
}

// The same for ForeignInvestment
```

I think this would help the implementation (and understanding) when things are simply passing downward

Reply:
I like this suggestion but I am not sure whether it's necessary once we merge `Investment` and `ForeignInvestment` traits in a subsequent PR? I would make the enum variants more descriptive though, i.e. `Invest` and `Redeem`.

Reply:
> necessary once we merge Investment and ForeignInvestment traits in a subsequent PR?

I would say this change is independent of merging those traits.

Agree with the enum variants 👍🏻